posthog/temporal/batch_exports/squash_person_overrides.py

```diff
@@ -161,6 +132,272 @@ class SerializablePersonOverrideToDelete(NamedTuple):
     oldest_event_at: str
 
 
+class PersonOverrideTuple(NamedTuple):
+    old_person_id: UUID
+    override_person_id: UUID
+
+
+class PostgresPersonOverridesManager:
+    def __init__(self, connection):
+        self.connection = connection
+
+    def fetchall(self, team_id: int) -> Sequence[PersonOverrideTuple]:
+        with self.connection.cursor() as cursor:
+            cursor.execute(
+                """
+                SELECT
+                    old_person.uuid,
+                    override_person.uuid
+                FROM posthog_personoverride override
+                LEFT OUTER JOIN
+                    posthog_personoverridemapping old_person
+                    ON override.team_id = old_person.team_id AND override.old_person_id = old_person.id
+                LEFT OUTER JOIN
+                    posthog_personoverridemapping override_person
+                    ON override.team_id = override_person.team_id AND override.override_person_id = override_person.id
+                WHERE override.team_id = %(team_id)s
+                """,
+                {"team_id": team_id},
+            )
+            return [PersonOverrideTuple(*row) for row in cursor.fetchall()]
+
+    def insert(self, team_id: int, override: PersonOverrideTuple) -> None:
+        with self.connection.cursor() as cursor:
+            mapping_ids = []
+            for person_uuid in (override.override_person_id, override.old_person_id):
+                cursor.execute(
+                    """
+                    INSERT INTO posthog_personoverridemapping(
+                        team_id,
+                        uuid
+                    )
+                    VALUES (
+                        %(team_id)s,
+                        %(uuid)s
+                    )
+                    ON CONFLICT("team_id", "uuid") DO NOTHING
+                    RETURNING id
+                    """,
+                    {"team_id": team_id, "uuid": person_uuid},
+                )
+                mapping_ids.append(cursor.fetchone())
+
+            cursor.execute(
+                """
+                INSERT INTO posthog_personoverride(
+                    team_id,
+                    old_person_id,
+                    override_person_id,
+                    oldest_event,
+                    version
+                )
+                VALUES (
+                    %(team_id)s,
+                    %(old_person_id)s,
+                    %(override_person_id)s,
+                    NOW(),
+                    1
+                );
+                """,
+                {
+                    "team_id": team_id,
+                    "old_person_id": mapping_ids[1],
+                    "override_person_id": mapping_ids[0],
+                },
+            )
+
+    def delete(self, person_override: SerializablePersonOverrideToDelete, dry_run: bool = False) -> None:
+        with self.connection.cursor() as cursor:
```

Comment:
Is this wrapping everything in a transaction? Not sure about the type of connection to check for myself, but thinking that wrapping the 3 queries that compose a delete in a transaction should be necessary to avoid any concurrency problems.

Reply:
It looks like by default `pyscopg2.connect` will open the connection in non-autocommit mode, so _all_ of [`delete_squashed_person_overrides_from_postgres`](https://github.com/PostHog/posthog/blob/5e177b9a5b4bb397c0cbc8428b8c2ac235af985c/posthog/temporal/batch_exports/squash_person_overrides.py#L681-L701) should be under transaction management.

Here's a kind of contrived isolated example to demonstrate and double check (it's been a while since I've used plain psycopg2), using the same connection string from the activity definition:

```python
(env) ted@revuelto posthog % DEBUG=1 DJANGO_SETTINGS_MODULE=posthog.settings python
Python 3.10.13 (main, Nov  2 2023, 15:29:51) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from django.conf import settings
>>> import psycopg2, psycopg2.extensions
>>> c = psycopg2.connect(dbname=settings.DATABASES["default"]["NAME"], user=settings.DATABASES["default"]["USER"], password=settings.DATABASES["default"]["PASSWORD"], host=settings.DATABASES["default"]["HOST"], port=settings.DATABASES["default"]["PORT"], **settings.DATABASES["default"].get("SSL_OPTIONS", {}))
2024-01-02 16:46:54 [warning  ] ['️Environment variable DEBUG is set - PostHog is running in DEVELOPMENT MODE!', 'Be sure to unset DEBUG if this is supposed to be a PRODUCTION ENVIRONMENT!']
ERROR: Unable to check license Apps aren't loaded yet.
>>> c.autocommit
False
>>> c.status == psycopg2.extensions.STATUS_READY
True
>>> cr = c.cursor()
>>> cr.execute("update django_content_type set app_label = app_label || '_test'")
>>> c.status == psycopg2.extensions.STATUS_IN_TRANSACTION
True
>>> c.rollback()
>>> c.status == psycopg2.extensions.STATUS_READY
True
```

I'm not actually sure if this is a _good_ thing (it seems like that could create locking issues between the override writer and squash, if they were to run concurrently) but it is transactionally safe, at least. 😄 

In any case, this path should now be unused (outside of tests) since I switched everything over to the flat overrides table today (which should be lock-free here), so I'm not too worried about the performance implications at this point as this path should be getting removed soon.

Reply:
Thanks for the showcase and explanation :+1:

I would insist that we were more explicit with our use of transactions instead of relying on psycopg's autocommit setting, but given that this is not in use and will be deleted I'm not too concerned.